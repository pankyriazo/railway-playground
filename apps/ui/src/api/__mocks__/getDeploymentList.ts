import { GetDeploymentListResponse } from "@/api/types";

export const mockGetDeploymentList = {
  data: {
    deployments: {
      edges: [
        {
          node: {
            id: "4b049de0-1ce0-4e81-8036-26b252e3285c",
            status: "SUCCESS",
            updatedAt: "2024-03-17T08:50:14.973Z",
            environmentId: "632da482-c5b1-4570-87af-9ea225c335cb",
          },
        },
        {
          node: {
            id: "38b68be4-f86d-4987-8c48-dee293fda404",
            status: "SUCCESS",
            updatedAt: "2024-03-16T19:41:25.747Z",
            environmentId: "1beb4373-6b1f-43cb-8150-0e7214acea82",
          },
        },
      ],
    },
  },
} as const satisfies GetDeploymentListResponse;

export const mockGetDeploymentListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetDeploymentListResponse;
