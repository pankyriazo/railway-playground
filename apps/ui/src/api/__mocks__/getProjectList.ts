import { GetProjectListResponse } from "@/api/types";

export const mockGetProjectList = {
  data: {
    me: {
      projects: {
        edges: [
          {
            node: {
              id: "a20a0d44-a6de-4804-acf1-58d436570fd7",
              name: "nauseating-cup",
              services: {
                edges: [
                  {
                    node: {
                      id: "0e62e7bf-2008-4452-bd32-e57cec3bb9a2",
                      name: "fascinated-motion",
                    },
                  },
                  {
                    node: {
                      id: "7ab42902-aae8-47b7-9976-81496dc61834",
                      name: "Postgres",
                    },
                  },
                ],
              },
              plugins: {
                edges: [],
              },
              environments: {
                edges: [
                  {
                    node: {
                      id: "1beb4373-6b1f-43cb-8150-0e7214acea82",
                      name: "production",
                    },
                  },
                  {
                    node: {
                      id: "632da482-c5b1-4570-87af-9ea225c335cb",
                      name: "staging",
                    },
                  },
                ],
              },
            },
          },
        ],
      },
    },
  },
} as const satisfies GetProjectListResponse;

export const mockGetProjectListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetProjectListResponse;
