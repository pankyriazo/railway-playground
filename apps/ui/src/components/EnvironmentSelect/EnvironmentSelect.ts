import { Component } from "@/types";
import { Edges, Node } from "@monorepo/types";

class EnvironmentSelect extends Component<HTMLSelectElement> {
  private _options: Edges<Node> = { edges: [] };
  private _selected: Node;
  private _onChange: (environment: Node) => void;

  constructor({
    options,
    selected,
    onChange,
  }: {
    options: Edges<Node>;
    selected: Node;
    onChange: (environment: Node) => void;
  }) {
    super();

    this._element = document.createElement("select");
    this._element.id = "environments";

    this._options = options;
    this._selected = selected;
    this._onChange = onChange;

    this._registerEvents();

    this.render();
  }

  private _registerEvents() {
    if (!this._element) {
      return () => {};
    }

    const handler = (event: Event) => {
      const target = event.target as HTMLSelectElement;

      const environment = this._options.edges.find(
        (edge) => edge.node.id === target.value
      );

      if (!environment) {
        return;
      }

      this._onChange(environment.node);
    };

    this._element.addEventListener("change", handler);

    return () => {
      this._element?.removeEventListener("change", handler);
    };
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.dataset.testid = "environment-select";

    for (const { node: environment } of this._options.edges) {
      const option = document.createElement("option");
      option.dataset.testid = "environment-select__option";
      option.value = environment.id;
      option.textContent = environment.name;
      this._element.appendChild(option);
    }

    this._element.value = this._selected.id;
  }
}

export default EnvironmentSelect;
