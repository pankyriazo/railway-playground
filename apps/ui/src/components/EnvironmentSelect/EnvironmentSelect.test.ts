import { render } from "@/test-helpers";
import { mockGetProjectList } from "@/api/__mocks__/getProjectList";
import EnvironmentSelect from "./EnvironmentSelect";

describe("EnvironmentSelect", () => {
  const onEnvironmentChange = jest.fn();

  beforeEach(() => {
    render(EnvironmentSelect, {
      options: mockGetProjectList.data.me.projects.edges[0].node.environments,
      selected:
        mockGetProjectList.data.me.projects.edges[0].node.environments.edges[0]
          .node,
      onChange: onEnvironmentChange,
    });
  });

  afterEach(() => {
    onEnvironmentChange.mockClear();
  });

  it("should render select element", () => {
    const select = document.querySelector("[data-testid=environment-select]");

    expect(select).not.toBeNull();
    expect(select).toMatchSnapshot();
  });

  it("should render options", () => {
    const options = document.querySelectorAll(
      "[data-testid=environment-select__option]"
    );

    expect(options).toHaveLength(2);
    expect(options[0]!.textContent).toBe("production");
    expect(options[1]!.textContent).toBe("staging");
  });
});
