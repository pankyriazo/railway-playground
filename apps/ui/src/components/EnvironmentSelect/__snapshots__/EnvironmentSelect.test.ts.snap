// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvironmentSelect should render select element 1`] = `
<select
  data-testid="environment-select"
  id="environments"
>
  <option
    data-testid="environment-select__option"
    value="1beb4373-6b1f-43cb-8150-0e7214acea82"
  >
    production
  </option>
  <option
    data-testid="environment-select__option"
    value="632da482-c5b1-4570-87af-9ea225c335cb"
  >
    staging
  </option>
</select>
`;
